### PR TITLE
Remove Tab for Referrers Page

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Remove Tab for Referrers Page
| Type?             | improvement
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes PrestaShop/PrestaShop#27064
| How to test?      | Cf. PrestaShop/PrestaShop#27064

### CODE REVIEW

**Review by NeOMakinG — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by khouloudbelguith — CHANGES_REQUESTED** [review]
Hi @Progi1984,

I have an issue following these steps:

1. Before the upgrade, go to BO > Shop Parameters > SEO & URLs & Refferes Tab
2. Add this page to the quick access image
1. Upgrade from PS1778 to this PR https://github.com/PrestaShop/PrestaShop/pull/27098 => Ok
3. Enable debug mode
4. Try to navigate to this link manually: `http://shop.com/admin-dev/index.php?controller=AdminReferrers&token` or via the quick access
5. See exception 
image

I tried a fresh install of this [PR](https://github.com/PrestaShop/PrestaShop/pull/27098), but I have a clear alert
image

I tried to compare it with the Module catalog also removed

- Upgrade from PS1778 to this [PR](https://github.com/PrestaShop/PrestaShop/pull/27098)
- Uninstall mbo module
- Try to navigate to this link manually: `http://shop.com/admin-dev/index.php?controller=AdminModulesCatalog`
- No exception was displayed 

image

Thanks!

**Comment by khouloudbelguith** [comment]
Hi @Progi1984,

I have another issue following these steps:
1. Install PS1778
2. Go to BO > Advanced Parameters > Team > Employees page
2. Create an employee with a `Salesman` profile 
3. Upgrade to this PR: https://github.com/PrestaShop/PrestaShop/pull/27098
4. Try to sign in with a new employee created
5. Go to BO > Shop Parameters > Traffic & Seo
6. See error 

image

After a fresh install => OK
image

Thanks!

**Comment by Progi1984** [comment]
@PierreRambaud @NeOMakinG Fixed last feedback from @khouloudbelguith

**Review by PierreRambaud — APPROVED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by khouloudbelguith — CHANGES_REQUESTED** [review]
Hi @Progi1984,

I tried two upgrades from PS1778 to develop branch & PS1784 to develop branch.
But I still have the same issue described [here](https://github.com/PrestaShop/autoupgrade/pull/466#pullrequestreview-900174867) & [here](https://github.com/PrestaShop/autoupgrade/pull/466#issuecomment-1059195865)
I attached a screen record

https://user-images.githubusercontent.com/16067358/156797605-0da00c3e-6af6-449f-8ce9-4d6781b7ee6b.mp4

PS: I can not remove the Waiting QA label.

Thanks!

**Review by PierreRambaud — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by khouloudbelguith — CHANGES_REQUESTED** [review]
Hi @Progi1984,

The second issue is fixed
But I still have the first issue following these steps:
1. Before the upgrade, go to BO > Shop Parameters > SEO & URLs & Refferes Tab
2. Add this page to the quick access image
1. Upgrade from PS1770 to develop => Ok
3. Enable debug mode
4. Try to navigate to this link manually: `http://shop.com/admin-dev/index.php?controller=AdminReferrers&token` or via the quick access
5. See exception 
image

I tried an upgrade from PS1770 to the develop branch using this PR of the autoupgrade.
I attached a screen record

https://user-images.githubusercontent.com/16067358/162018464-0ca70bb7-cb2d-4f6e-ad5f-b7f37f715b6b.mp4

Thanks!

**Review by jolelievre — APPROVED** [review]
Thanks @Progi1984

**Review by jolelievre — CHANGES_REQUESTED** [review]
Oh actually seems like some bug was found

**Comment by Progi1984** [comment]
Found the bug. Thanks @atomiix for the help.

**Review by jolelievre — APPROVED** [review]

**Review by MeKeyCool — DISMISSED** [review]

**Review by atomiix — CHANGES_REQUESTED** [review]

File: classes/TaskRunner/Upgrade/UpgradeFiles.php [hunk 1]
```diff
@@ -252,8 +252,8 @@ protected function warmUp()
             foreach ($list_files_diff as $k => $path) {
                 if (preg_match('#autoupgrade#', $path)) {
                     unset($list_files_diff[$k]);
-                } else {
-                    $list_files_diff[$k] = str_replace('/' . 'admin', '/' . $admin_dir, $path);
+                } elseif (strpos($path, '/admin') === 0) {
```
[the remark this answers is not in the document]
Comment (Progi1984) [reply classes/TaskRunner/Upgrade/UpgradeFiles.php:255]: @atomiix Euh... No. That updates only path who begins with `/admin`. Please read again.

[the remark this answers is not in the document]
Comment (atomiix) [reply classes/TaskRunner/Upgrade/UpgradeFiles.php:255]: The comment is clear: `// only keep list of files to delete. The modified files will be listed with _listFilesToUpgrade`. We are adding the files to delete. And even if it was the file to update, it should not be **only** files that starts with `/admin`, it should be all the files. The only thing added, is that if a file starts with `/admin` we replace its path with the actual `admin<something>` path.

[the remark this answers is not in the document]
Comment (Progi1984) [reply classes/TaskRunner/Upgrade/UpgradeFiles.php:255]: Victory :p

[the remark this answers is not in the document]
Comment (atomiix) [reply classes/TaskRunner/Upgrade/UpgradeFiles.php:255]: Riiiiight! Got it now 😄

**Review by atomiix — APPROVED** [review]

**Review by khouloudbelguith — COMMENTED** [review]
Hi @Progi1984,

I have a difference between the fresh install develop branch and PS8.0.0 upgraded from PS1778 using this PR

Steps to reproduce the behavior:
1. Install PS1778
2. Add those pages to the quick access management:
- index.php?controller=AdminReferrers
- index.php?controller=AdminReferrers&addreferrer
- index.php?controller=AdminReferrers&viewreferrer=&id_referrer=1
- index.php?controller=AdminReferrers&updatereferrer=&id_referrer=1
3. Create an employee with a salesman profile
4. Upgrade to the develop branch using this PR
5. Go to the database > search with this expression `Referrers`

Check the table `sk_authorization_role` 
image

Check the table `sk_configuration`
image


I tried Fresh install develop
image

PS: I tried this issue: https://github.com/PrestaShop/autoupgrade/pull/466#pullrequestreview-933744064 and it is fixed :tada: 

Thanks!

**Comment by Progi1984** [comment]
@khouloudbelguith Fixed. Thanks

**Comment by Progi1984** [comment]
Ping @jolelievre & @atomiix for review :)

**Review by matks — COMMENTED** [review]
Suggestion: if we need this logic in the future it would be nice to be able to reuse it.

There is a function `register_tab()` in https://github.com/PrestaShop/autoupgrade/blob/dev/upgrade/php/add_new_tab.php

Maybe we should create a function `unregister_tab()`

**Review by atomiix — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]
Thanks @Progi1984 

@matks regarding your suggestion to have an `unregister_tab` function I agree it could come in handy I guess next time we need to do this we can refacto this code and create the function.

**Comment by MatShir** [comment]
When the core removes a page, we should remove the quick access linked to this page.

**Comment by Progi1984** [comment]
Thanks @khouloudbelguith & @MatShir : I remove Quick Access in the upgrade.

@PrestaShop/prestashop-maintainers Ready for review (Sorry for the ping, but there is the label "Waiting for author")

**Review by matthieu-rolland — APPROVED** [review]

File: upgrade/sql/8.0.0.sql [hunk 2]
```diff
@@ -5,6 +5,29 @@ DROP TABLE IF EXISTS `PREFIX_referrer`;
...
+DELETE FROM `PREFIX_configuration`
+  WHERE `name` IN ('PS_REFERRERS_CACHE_LIKE', 'PS_REFERRERS_CACHE_DATE');
+## Remove Quick Access
+DELETE FROM `PREFIX_quick_access_lang`
+  WHERE id_quick_access IN (
+    SELECT id_quick_access FROM `PREFIX_quick_access` WHERE link LIKE '%controller=AdminReferrers%'
+  );
+DELETE FROM `PREFIX_quick_access`
+  WHERE link LIKE '%controller=AdminReferrers%';
+
```
Comment (jolelievre) [inline upgrade/sql/8.0.0.sql:30]: Similarly to what was done in the previous request maybe we could generalize this with:
```sql
DELETE FROM `PREFIX_quick_access`
  WHERE link LIKE '%controller=AdminReferrers%';
DELETE FROM `PREFIX_quick_access_lang` WHERE `id_quick_access` NOT IN (SELECT id_quick_access FROM `PREFIX_quick_access`);
```

**Review by jolelievre — APPROVED** [review]
Hi @Progi1984 

I added a suggestion to generalize one of the query, mostly because I think all these requests could be moved into a function that could be reused later in the future.

Except the query related to SalesMan profile which is quite specific all the others seem top be generic for any other Tab we would need to remove. And maybe even this one could be done actually, because we could check for the removed Tab if its parent still has some children left (for a specific profile) if not then it can be purged from the DB no?

I'll still approve the PR because this is a suggestion for improvement and the PR already works this way, but still it would be a nice tool for future versions.

(Edit: the function idea is similar to what @matthieu-rolland did in this PR for tab renaming https://github.com/PrestaShop/autoupgrade/pull/489)

**Comment by AureRita** [comment]
Hi @Progi1984 ,

After some test, I tried to upgrade to the PR or to the 8.0.0 and I've always the "error 500" : 
image

It doesn't happens when I do an upgrade from 1.7.7.8 to 1.7.8.6 by a local archive

**Comment by AureRita** [comment]
My fault, I didn't uninstall the different module before the installation of 8.0.0

**Comment by AureRita** [comment]
Hi @Progi1984 , 

I tested your PR and you've corrected the bug with the database 

you also remove the quick access when we have an update, as we can see on this video : 

https://user-images.githubusercontent.com/100699445/179553005-9d261cda-05d8-498c-bab6-f5c0fe5f90bc.mp4

but, unfortunately, we always have the exceptions showed when we go on the differents webpages by a bookmark (from firefox) or if we copy/paste the link of the differents pages, as we can see on this video 

https://user-images.githubusercontent.com/100699445/179553510-5216f85d-c2b4-4c34-8057-1b287c907efc.mp4

the different page are : 
index.php?controller=AdminReferrers
index.php?controller=AdminReferrers&addreferrer
index.php?controller=AdminReferrers&viewreferrer=&id_referrer=1
index.php?controller=AdminReferrers&updatereferrer=&id_referrer=1

When we do a fresh install of the develop version, we have this intended behaviour : 
image

**Comment by Progi1984** [comment]
@AureRita The final behavior is the expected behavior. The page doesn't exist like this one :

image

**Review by AureRita — APPROVED** [review]
Hi @Progi1984 

After many investigation, your PR is, indeed,  QA validated, 

Thank you for all your modification

File: upgrade/sql/8.0.0.sql (hunk 2 above)
Comment (matks) [reply upgrade/sql/8.0.0.sql:30]: Finally method will be implemented in https://github.com/PrestaShop/autoupgrade/pull/504/ 🎉

**Pull request merged**